### ISSUE TITLE
Issue when trying to square up the expense

### ISSUE THREAD

**Issue by therealharpaljadeja** [celo-org/ethglobal-istanbul#19]
The 2-step process of approving cUSD and calling a function that transfers cUSD is not working/unreliable in the Site Tester inside MiniPay

The frontend code

https://github.com/celo-org/ethglobal-istanbul/blob/c7a16ff9d9ef86dfaa05b7761c6211b753b7d41c/ethglobal-istanbul/code/packages/react-app/components/Settle.tsx#L30-L91

The smart contract code

https://github.com/celo-org/ethglobal-istanbul/blob/c7a16ff9d9ef86dfaa05b7761c6211b753b7d41c/ethglobal-istanbul/code/packages/hardhat/contracts/SplitPay.sol#L125-L157

**Comment by therealharpaljadeja** [comment]
The deployed version of the dApp is [here](https://splitpay-minipay.vercel.app/), that you can try inside MiniPay.

1. Using the "Add Expense" button, you can add an expense that you want to split among other addresses.

<img width="601" alt="image" src="https://github.com/celo-org/ethglobal-istanbul/assets/38040789/bf31ce9f-c1ef-41ab-86f2-3fd647932ec3">

2. Using the "Add Friend" you can paste a wallet address or type a phone number registered under MiniPay issue to get the address and then complete adding the expense.

<img width="542" alt="image" src="https://github.com/celo-org/ethglobal-istanbul/assets/38040789/310838d3-b1eb-4272-9a26-f171914ca8a4">

3. Lastly when you connect the other wallet, you will see the payments that need to be squared up.

<img width="566" alt="image" src="https://github.com/celo-org/ethglobal-istanbul/assets/38040789/61252262-2a73-4f1f-9d80-0ac1fb07adc3">

**Comment by therealhampus** [comment]
The frontend looks like what I imagined it would. Albeit I haven't worked with 'wagmi' before so I'm not entirely sure what all calls do internally.

What kind of errors are you receiving from calling `publicClient.waitForTransactionReceipt({  hash });`?

While debugging, I can't see any errors in the console.

**Comment by therealhampus** [comment]
internal error found in logs.

gas estimation request:
```
{"method":"eth_estimateGas","params":[{"from":"0x8abe5bee4f3768f90acd134792b07eef7ef50711","gasPrice":"0x16930cd62","value":"0x0","data":"0x66aeccbf000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000006f05b59d3b2000000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000644696e6e6572000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000008abe5bee4f3768f90acd134792b07eef7ef50711000000000000000000000000219a8442870b359a363902578f25042473af550e","to":"0x36eef317f736fc5d7d0caae80a1bd7ad1d93b874","feecurrency":"0x874069fa1eb16d44d622f2e0ca25eea172369bc1"}],"jsonrpc":"2.0","id":1}
```
response:
```
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"execution reverted"}}
```

**Comment by therealharpaljadeja** [comment]
> The frontend looks like what I imagined it would. Albeit I haven't worked with 'wagmi' before so I'm not entirely sure what all calls do internally.
> 
> What kind of errors are you receiving from calling `publicClient.waitForTransactionReceipt({ hash });`?
> 
> While debugging, I can't see any errors in the console.

How did you get access to the console? Are you trying in a browser?

**Comment by therealhampus** [comment]
> How did you get access to the console? Are you trying in a browser?
tried inspecting the device via `chrome://inspect/#devices`

Although from my latest reply, I'm wondering if you are actually receiving and error from us?

**Comment by therealhampus** [comment]
Also, what support is there for test net? I'm on alfajores when I'm testing this.

**Comment by therealharpaljadeja** [comment]
> Also, what support is there for test net? I'm on alfajores when I'm testing this.

I see why you are getting the error 😂, this is on Mainnet

**Comment by therealhampus** [comment]
That explains the initial hurdle hehe.

I'll switch net and see what I get

**Comment by therealhampus** [comment]
Switching to mainnet I'm able to create an expense although confirmation time was upwards of a minute.

[here's the tx](https://explorer.celo.org/mainnet/tx/0xb8f1363b7ab169b7210bb1307623601693ec16efbd05c3408233fe4bcf45272d)

I can't see anything out of order really.

**Comment by therealhampus** [comment]
Approval check is taking several minutes (still going) on the "payer" side of things. Known? seems that whatever node `publicClient` is talking to it's really slow.

**Comment by therealharpaljadeja** [comment]
> Approval check is taking several minutes (still going) on the "payer" side of things. Known? seems that whatever node `publicClient` is talking to it's really slow.

Yes, Approval is taking forever that is the issue, I was never actually able to go through. Once the approval is done there is one more step which is what the user is not able to perform because the approval never ends and the UI to square up or settle up never renders.

**Comment by therealhampus** [comment]
> Yes, Approval is taking forever that is the issue
Just to clarify: do you mean that waiting for the receipt of an approval contract call is slow, or the call to check approval?

OT:
So could it be the node being slow to get the receipt, or do you suspect that there's another issue?

The other thing that I see as a possible explanation for this is gas. What I mean is that it might be too low and it ends up not being prioritized. However, there's no internal difference between a web3 transaction and a MiniPay send transaction. And the latter is blazing fast, so I don't see why web3 requests wouldn't be for your dApp.

**Comment by therealharpaljadeja** [comment]
> > Yes, Approval is taking forever that is the issue
> 
> Just to clarify: do you mean that waiting for the receipt of an approval contract call is slow, or the call to check approval?
> 
> OT: So could it be the node being slow to get the receipt, or do you suspect that there's another issue?
> 
> The other thing that I see as a possible explanation for this is gas. What I mean is that it might be too low and it ends up not being prioritized. However, there's no internal difference between a web3 transaction and a MiniPay send transaction. And the latter is blazing fast, so I don't see why web3 requests wouldn't be for your dApp.

I think all calls in general are slow because even the read call to get expenses and pending settlements is slow, one thing I can think of is that internally using wagmi I might not be interacting with `window.ethereum`.

**Comment by therealhampus** [comment]
> Let me test the following
> 
> * [ ]  Read calls only (Getting effective balance, expenses and pending payments)
> * [ ]  Check if approval required
> * [ ]  Send transaction directly using `window.ethereum` and not using wagmi
> * [ ]  Wait for transaction receipt using `window.ethereum`

Yes please do that and see if there's better response from MiniPay.

> If the above works fast then internally wagmi is not interacting with `window.ethereum` or something strange is happening.
> 
> As I write this I also wonder how am I getting the Minipay prompt to do cUSD approval, that means `window.ethereum` is being used.

MiniPay's injected provider only attaches itself to `window.ethereum`. It used to be available on `window.web3` for legacy reasons, but it was removed.

**Comment by therealharpaljadeja** [comment]
@therealhampus 

I replaced the `publicClient` with `window.ethereum` but the issues are still the same

```js
let provider = new ethers.providers.Web3Provider(
    window.ethereum
);

let splitPayContract = new ethers.Contract(
    SPLITPAY_CONTRACT_ADDRESS,
    SplitPayAbi,
    provider
);

let { effectiveBalance, expenses, settlements } =
    await splitPayContract.getUserStats(address);
```


I can make the first read call but then the next read call right after the first one doesn't work, so only the first read call works and the first transaction call works. If I do anything one after the other then it doesn't work for me.

**Comment by therealhampus** [comment]
@therealharpaljadeja

I see. I'll dig in to it and see what I can find. on our end.

Some clarification for my own sake: Do the read calls trigger UI in MiniPay that the user should sign?  Or are the calls just made from the frontend?

**Comment by therealharpaljadeja** [comment]
No read calls don't trigger any UI, and no signing is required.

**Comment by therealhampus** [comment]
Thanks, If these changes are live in you app I can investigate what happens on our end.

**Comment by therealharpaljadeja** [comment]
yes [here](https://splitpay-minipay.vercel.app)

**Comment by therealhampus** [comment]
I don't see the text either.

So the first 30 or so requests from split are not interesting (just a bunch of address requests and a couple of chain requests).

Then there's two requests that I can't decode (with responses):

```
>>> 09:14:02.791: {"method":"eth_call","params":[{"to":"0x36eef317f736fc5d7d0caae80a1bd7ad1d93b874","data":"0x4e43603a0000000000000000000000008abe5bee4f3768f90acd134792b07eef7ef50711"},"latest"]}

<<< 09:14:02.920: {"jsonrpc":"2.0","id":38,"result":"0x0000000000000000000000000000000000000000000000000000000000000020fffffffffffffffffffffffffffffffffffffffffffffffffc87d25316270000000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000b"}
```
```
>>> 09:14:05.041: {"method":"eth_call","params":[{"to":"0x36eef317f736fc5d7d0caae80a1bd7ad1d93b874","data":"0x735e5db60000000000000000000000000000000000000000000000000000000000000006"},"latest"]}

<<< 09:14:05.269: {"jsonrpc":"2.0","id":41,"result":"0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000006f05b59d3b2000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000008abe5bee4f3768f90acd134792b07eef7ef50711000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000b000000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000000644696e6e65720000000000000000000000000000000000000000000000000000"}
```

What information exactly are you getting with these calls? I guess at least one would be the effective balance?

These request/response pairs are the last ones that MiniPay handles, without clicking anything on the webpage. Clicking on "square it up" does not trigger and subsequent requests.

**Comment by therealhampus** [comment]
From the account that needs to settle, I see the following:

```
>>> 10:17:01.103: {"method":"eth_call","params":[{"to":"0x36eef317f736fc5d7d0caae80a1bd7ad1d93b874","data":"0x4e43603a000000000000000000000000219a8442870b359a363902578f25042473af550e"},"latest"]}

<<< 10:17:01.221: {"jsonrpc":"2.0","id":30,"result":"0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000003782dace9d900000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000c"}
```

and then there's another pair a full minute later on:

```
>>> 10:18:01.767: {"method":"eth_call","params":[{"to":"0x36eef317f736fc5d7d0caae80a1bd7ad1d93b874","data":"0x4c582380000000000000000000000000000000000000000000000000000000000000000c"},"latest"]}

<<< 10:18:01.904: {"jsonrpc":"2.0","id":34,"result":"0x000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000003782dace9d90000000000000000000000000000219a8442870b359a363902578f25042473af550e0000000000000000000000000000000000000000000000000000000000000000"}
```

The latter is yet another contract call that MiniPay doesn't decode.


Again, there's nothing on the "square it up" page. And there's no calls incoming after the last request/response pair.

**Comment by therealhampus** [comment]
> "Square it up" tab gets its content from the second call.
Okay. Is `0x735e5db6` and `0x4c582380` perhaps mapping to `getExpense` and `getSettlement` respectively?

Anyhow, this content is actually fetched and returned quite fast to the frontend. But there's still no info under "Expenses" nor "Square it up"...

**Comment by therealhampus** [comment]
And both these calls are considered read calls correct?

**Comment by therealharpaljadeja** [comment]
> > "Square it up" tab gets its content from the second call.
> > Okay. Is `0x735e5db6` and `0x4c582380` perhaps mapping to `getExpense` and `getSettlement` respectively?
> 
> Anyhow, this content is actually fetched and returned quite fast to the frontend. But there's still no info under "Expenses" nor "Square it up"...

I guess it's unreliable, I used the app on an Android phone and it showed me the "No Expenses to settle" text, but looking up address for a phone number didn't work 😅

> And both these calls are considered read calls correct?

Yes

**Comment by therealhampus** [comment]
So I downloaded the repo and tried running the code locally to see what may have been going on.

Have you noticed the error from @wagmi:

```
client.js:1 ConnectorAlreadyConnectedError: Connector already connected
    at connect (chunk-TSH6VVF4.js:654:1)
    at Object.mutationFn (index.js:722:17)
    at Object.fn (mutation.mjs:87:1)
    at run (retryer.mjs:101:1)
    at createRetryer (retryer.mjs:149:1)
    at executeMutation (mutation.mjs:81:35)
    at Mutation.execute (mutation.mjs:138:1)
```

Seems that it's thrown from Header.tsx where the dapp tries to connect if it finds MiniPay.

What I found is that it connects successfully, meaning `useAccounts` from wagmi should work in other places in the code. But the connect calls never stop and it's severely slowing down the app. I modified the `useEffect` to only try connect if the status is `idle`:

```
    const { connect, status } = useConnect({
        connector: new InjectedConnector(),
        onSettled(data, error) {
            console.log('Settled data:', data);
            console.log("Settled err:", error);
        
        },
    });

    useEffect(() => {
        console.log(status);
        if (window.ethereum && window.ethereum.isMiniPay && status === "idle") {
            setHideConnectBtn(true);
            console.log('connecting...');
            connect();
        }
    }, [connect, status]);
```

Mind you that it won't really handle connection errors. 

IMO the app is smoother on my device after this change. And I also tend to get all the info I would expect in each tab (expense and square it up).


What do you make of this?

**Comment by therealharpaljadeja** [comment]
It worked! 

I think yes, the `useEffect` must be going into an infinite connection loop which made the app slow.

I am not able to figure out, how did you inspect the app opened inside Site Tester because `chrome://inspect/#devices` is not letting me inspect my app running inside Site Tester

**Comment by therealhampus** [comment]
Great!

> I am not able to figure out, how did you inspect the app opened inside Site Tester because chrome://inspect/#devices is not letting me inspect my app running inside Site Tester

I reckon that you're running MiniPay inside of Opera Mini?

**Comment by therealharpaljadeja** [comment]
Yes

I guess I can't see logs 

<img width="1235" alt="image" src="https://github.com/celo-org/ethglobal-istanbul/assets/38040789/9c542b24-82ff-43d9-804c-77e6e8440c7c">

**Comment by therealhampus** [comment]
> I guess I can't see logs
>
> Dev mode is on.
I'll raise this with the team so we can investigate why it doesn't seem to work.

**Comment by therealharpaljadeja** [comment]
Now that error is solved I am getting another which is this...

I have tested doing transactions using other wallets and in the browser and it works but not in MiniPay.

<img width="540" alt="Screenshot 2023-11-16 at 2 29 34 PM" src="https://github.com/celo-org/ethglobal-istanbul/assets/38040789/3c47ac02-2f46-4969-b0c1-dbcc0669b5fe">

**Comment by therealhampus** [comment]
When does this error occur? When adding an expense, or when you settle one?

**Comment by therealharpaljadeja** [comment]
For every contract interaction after the fingerprint touch

**Comment by therealhampus** [comment]
Let me try and see what could be the issue here.

**Comment by therealhampus** [comment]
What version of MiniPay are you on? I've been making expenses and squared them without issues.

We may have done unreleased changes that fixed this issue. If you can report the version number in minipay (settings -> about -> version) I can examine recent changes and try from that release.

**Comment by therealhampus** [comment]
Also, there's practically _nothing_ about this decryption error when searching for some explanation about it...

I reckon that you see if after the transaction has been signed by miniPay, and you start waiting for the receipt (with the returned tx) in your dApp. It would make sense to look at the tx on an explorer. Do you have the wallet address for which this tx fails? We could see if the explorer reveals something about the transaction.

**Comment by therealharpaljadeja** [comment]
> What version of MiniPay are you on? I've been making expenses and squared them without issues.
> 
> We may have done unreleased changes that fixed this issue. If you can report the version number in minipay (settings -> about -> version) I can examine recent changes and try from that release.

Version: 2023.42.449475



> Also, there's practically _nothing_ about this decryption error when searching for some explanation about it...
> 
> I reckon that you see if after the transaction has been signed by miniPay, and you start waiting for the receipt (with the returned tx) in your dApp. It would make sense to look at the tx on an explorer. Do you have the wallet address for which this tx fails? We could see if the explorer reveals something about the transaction.

The address is: `0xAF4411233Fc9e8b384b9008E8aac5f0ba810d27a`

[Here](https://celoscan.io/address/0xAF4411233Fc9e8b384b9008E8aac5f0ba810d27a) is the link to the explorer. I think the transaction is rejected before it is even submitted to the blockchain.

The address of the dApp contract on Mainnet - `0x36Eef317F736FC5d7D0CAAe80a1bD7aD1D93B874`

[Here](https://celoscan.io/address/0x36Eef317F736FC5d7D0CAAe80a1bD7aD1D93B874) is the link to the dApp contract on the explorer, no transactions have been submitted by the above wallet address.

**Comment by therealhampus** [comment]
> The address is: `0xAF4411233Fc9e8b384b9008E8aac5f0ba810d27a`
> 
> [Here](https://celoscan.io/address/0xAF4411233Fc9e8b384b9008E8aac5f0ba810d27a) is the link to the explorer. I think the transaction is rejected before it is even submitted to the blockchain.

Yeah it looks like it. So something goes wrong either inside of the viem library, or when the transaction is submitted. Have you tried to build your app with source maps on? It's a shame that I can't reproduce this issue since I would've gotten the stacktrace for this issue.

> The address of the dApp contract on Mainnet - `0x36Eef317F736FC5d7D0CAAe80a1bD7aD1D93B874`
> 
> [Here](https://celoscan.io/address/0x36Eef317F736FC5d7D0CAAe80a1bD7aD1D93B874) is the link to the dApp contract on the explorer, no transactions have been submitted by the above wallet address.
> 
> Could it be because MiniPay adds `feeCurrency` property to the transaction?

We do add the feeCurrency, if the dApp hasn't specified the property itself. But shouldn't `wagmi` be fully capable of handling that property? Or are you using something else after we started this discussion?

**Comment by therealhampus** [comment]
> Could it be because MiniPay adds `feeCurrency` property to the transaction?

I shouldn't matter, since the call you're making is asking MiniPay to do all of the transaction signing and sending.

I can't find a single piece of information online that shows viem throwing `TransactionExecutionError` with `Details: Decryption error`. From their source code I do see that they get `Details` from a rejected RPC request. Meaning the error is likely thrown and caught entirely in MiniPay. And then it sends back a JSON RPC error response with `Decryption error` as its response message.

I have some things for you to try:

-  If you still encounter the same problem. Try to send a transaction in the MiniPay app, using the send functionality from the app's main screen. It is likely that it will fail too.
- Try restarting the emulator device.

**Comment by therealharpaljadeja** [comment]
Closed by #29
